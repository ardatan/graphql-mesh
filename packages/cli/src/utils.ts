export const stripJSONComments = (data: string) => {
  const re = /\/\/(.*)/g;
  const re2 = /\/\*(.*)\*\//g;
  return data.replace(re, '').replace(re2, '');
};
