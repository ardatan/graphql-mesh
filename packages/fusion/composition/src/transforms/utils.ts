export class TransformValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'TransformValidationError';
  }
}
