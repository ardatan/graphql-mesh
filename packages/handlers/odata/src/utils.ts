export function getUrlString(url: URL) {
  return decodeURIComponent(url.toString()).split('+').join(' ');
}

export function addIdentifierToUrl(url: URL, identifierFieldName: string, identifierFieldTypeRef: string, args: any) {
  url.href += `/${args[identifierFieldName]}/`;
}
