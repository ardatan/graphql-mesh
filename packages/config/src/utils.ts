import { parse } from 'graphql';
import { MeshHandlerLibrary, KeyValueCache, YamlConfig, MergerFn, ImportFn, MeshPubSub } from '@graphql-mesh/types';
import { resolve } from 'path';
import { IResolvers, printSchemaWithDirectives } from '@graphql-tools/utils';
import { paramCase } from 'param-case';
import { loadTypedefs } from '@graphql-tools/load';
import { GraphQLFileLoader } from '@graphql-tools/graphql-file-loader';
import { get, set, kebabCase } from 'lodash';
import { stringInterpolator } from '@graphql-mesh/utils';
import { mergeResolvers } from '@graphql-tools/merge';
import { PubSub, withFilter } from 'graphql-subscriptions';
import { EventEmitter } from 'events';
import { CodeFileLoader } from '@graphql-tools/code-file-loader';
import StitchingMerger from '@graphql-mesh/merger-stitching';

export async function getPackage<T>(name: string, type: string, importFn: ImportFn): Promise<T> {
  const casedName = paramCase(name);
  const casedType = paramCase(type);
  const possibleNames = [
    `@graphql-mesh/${casedName}`,
    `@graphql-mesh/${casedName}-${casedType}`,
    `@graphql-mesh/${casedType}-${casedName}`,
    casedName,
    `${casedName}-${casedType}`,
    `${casedType}-${casedName}`,
    casedType,
  ];
  if (name.includes('-')) {
    possibleNames.push(name);
  }
  const possibleModules = possibleNames.concat(resolve(process.cwd(), name));

  for (const moduleName of possibleModules) {
    try {
      const exported = await importFn(moduleName);

      return (exported.default || exported.parser || exported) as T;
    } catch (err) {
      if (
        !err.message.includes(`Cannot find module '${moduleName}'`) &&
        !err.message.includes(`Could not locate module`)
      ) {
        throw new Error(`Unable to load ${type} matching ${name}: ${err.message}`);
      }
    }
  }

  throw new Error(`Unable to find ${type} matching ${name}`);
}

export async function getHandler(name: keyof YamlConfig.Handler, importFn: ImportFn): Promise<MeshHandlerLibrary> {
  const handlerFn = await getPackage<MeshHandlerLibrary>(name.toString(), 'handler', importFn);

  return handlerFn;
}

export async function resolveAdditionalTypeDefs(baseDir: string, additionalTypeDefs: string) {
  if (additionalTypeDefs) {
    const sources = await loadTypedefs(additionalTypeDefs, {
      cwd: baseDir,
      loaders: [new CodeFileLoader(), new GraphQLFileLoader()],
    });
    return sources.map(source => source.document || parse(source.rawSDL || printSchemaWithDirectives(source.schema)));
  }
  return undefined;
}

export async function resolveAdditionalResolvers(
  baseDir: string,
  additionalResolvers: (
    | string
    | YamlConfig.AdditionalStitchingResolverObject
    | YamlConfig.AdditionalSubscriptionObject
  )[],
  importFn: ImportFn,
  pubsub: MeshPubSub
): Promise<IResolvers> {
  const loadedResolvers = await Promise.all(
    (additionalResolvers || []).map(async additionalResolver => {
      if (typeof additionalResolver === 'string') {
        const filePath = additionalResolver;

        const exported = await importFn(resolve(baseDir, filePath));
        let resolvers = null;

        if (exported.default) {
          if (exported.default.resolvers) {
            resolvers = exported.default.resolvers;
          } else if (typeof exported.default === 'object') {
            resolvers = exported.default;
          }
        } else if (exported.resolvers) {
          resolvers = exported.resolvers;
        }

        if (!resolvers) {
          console.warn(`Unable to load resolvers from file: ${filePath}`);

          return {};
        }

        return resolvers;
      } else {
        if ('pubsubTopic' in additionalResolver) {
          return {
            [additionalResolver.type]: {
              [additionalResolver.field]: {
                subscribe: withFilter(
                  (root, args, context, info) => {
                    const resolverData = { root, args, context, info };
                    const topic = stringInterpolator.parse(additionalResolver.pubsubTopic, resolverData);
                    return pubsub.asyncIterator(topic);
                  },
                  (root, args, context, info) => {
                    return additionalResolver.filterBy ? eval(additionalResolver.filterBy) : true;
                  }
                ),
                resolve: (payload: any) => {
                  if (additionalResolver.returnData) {
                    return get(payload, additionalResolver.returnData);
                  }
                  return payload;
                },
              },
            },
          };
        } else {
          return {
            [additionalResolver.type]: {
              [additionalResolver.field]: {
                selectionSet: additionalResolver.requiredSelectionSet,
                resolve: async (root: any, args: any, context: any, info: any) => {
                  const resolverData = { root, args, context, info };
                  const methodArgs: any = {};
                  for (const argPath in additionalResolver.args) {
                    set(methodArgs, argPath, stringInterpolator.parse(additionalResolver.args[argPath], resolverData));
                  }
                  const result = await context[additionalResolver.targetSource].api[additionalResolver.targetMethod](
                    methodArgs,
                    {
                      selectedFields: additionalResolver.resultSelectedFields,
                      selectionSet: additionalResolver.resultSelectionSet,
                      depth: additionalResolver.resultDepth,
                    }
                  );
                  return additionalResolver.returnData ? get(result, additionalResolver.returnData) : result;
                },
              },
            },
          };
        }
      }
    })
  );

  return mergeResolvers(loadedResolvers);
}

export async function resolveCache(
  cacheConfig: YamlConfig.Config['cache'],
  importFn: ImportFn
): Promise<KeyValueCache | undefined> {
  if (cacheConfig) {
    const cacheName = Object.keys(cacheConfig)[0];
    const config = cacheConfig[cacheName];

    const moduleName = kebabCase(cacheName.toString());
    const pkg = await getPackage<any>(moduleName, 'cache', importFn);
    const Cache = pkg.default || pkg;

    return new Cache(config);
  }
  const InMemoryLRUCache = await import('@graphql-mesh/cache-inmemory-lru').then(m => m.default);
  const cache = new InMemoryLRUCache();
  return cache;
}

export async function resolvePubSub(
  pubsubYamlConfig: YamlConfig.Config['pubsub'],
  importFn: ImportFn
): Promise<MeshPubSub> {
  if (pubsubYamlConfig) {
    let pubsubName: string;
    let pubsubConfig: any;
    if (typeof pubsubYamlConfig === 'string') {
      pubsubName = pubsubYamlConfig;
    } else {
      pubsubName = pubsubYamlConfig.name;
      pubsubConfig = pubsubYamlConfig.config;
    }

    const moduleName = kebabCase(pubsubName.toString());
    const pkg = await getPackage<any>(moduleName, 'pubsub', importFn);
    const PubSub = pkg.default || pkg;

    return new PubSub(pubsubConfig);
  } else {
    const eventEmitter = new EventEmitter({ captureRejections: true });
    eventEmitter.setMaxListeners(Infinity);
    const pubsub = new PubSub({ eventEmitter }) as MeshPubSub;

    return pubsub;
  }
}

export async function resolveMerger(mergerConfig: YamlConfig.Config['merger'], importFn: ImportFn): Promise<MergerFn> {
  if (mergerConfig) {
    const pkg = await getPackage<any>(mergerConfig, 'merger', importFn);
    return pkg.default || pkg;
  }
  return StitchingMerger;
}
