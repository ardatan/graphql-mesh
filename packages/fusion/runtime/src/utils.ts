import { print, type DocumentNode, type ExecutionResult, type GraphQLSchema } from 'graphql';
import type { TransportOptions } from '@graphql-mesh/serve-runtime';
import type {
  Transport,
  TransportContext,
  TransportEntry,
  TransportGetSubgraphExecutor,
  TransportGetSubgraphExecutorOptions,
} from '@graphql-mesh/transport-common';
import type { Logger } from '@graphql-mesh/types';
import { iterateAsync, mapMaybePromise } from '@graphql-mesh/utils';
import {
  isAsyncIterable,
  isDocumentNode,
  mapAsyncIterator,
  printSchemaWithDirectives,
  type ExecutionRequest,
  type Executor,
  type Maybe,
  type MaybePromise,
} from '@graphql-tools/utils';

export type { TransportEntry, TransportGetSubgraphExecutor, TransportGetSubgraphExecutorOptions };

export type Transports =
  | {
      [Kind in string]: MaybePromise<
        | Transport<Kind> // named export
        | {
            default: Transport<Kind>; // property on default export
          }
      >;
    }
  | (<Kind extends string>(
      kind: Kind,
    ) => MaybePromise<
      | Transport<Kind> // named export
      | {
          default: Transport<Kind>; // property on default export
        }
    >);

export type TransportExecutorFactoryGetter = (
  kind: string,
) => MaybePromise<TransportGetSubgraphExecutor>;

function createTransportExecutorFactoryGetter(
  transports: Transports,
  transportOptions: TransportOptions,
  logger: Logger,
): TransportExecutorFactoryGetter {
  return async function getTransport(kind) {
    let transportGetSubgraphExecutor: TransportGetSubgraphExecutor;
    let transport =
      typeof transports === 'function' ? await transports(kind) : await transports?.[kind];
    if (transport) {
      transportGetSubgraphExecutor =
        'default' in transport
          ? transport.default.getSubgraphExecutor
          : transport.getSubgraphExecutor;
    } else {
      const childLogger = logger.child?.(kind);
      try {
        transport = await import(`@graphql-mesh/transport-${kind}`);
      } catch (err) {
        childLogger.error(err);
        throw new Error(
          `No transport found for ${kind}. Please make sure you have installed @graphql-mesh/transport-${kind} or defined the transport config in "mesh.config.ts"`,
        );
      }
      transportGetSubgraphExecutor =
        'default' in transport
          ? transport.default.getSubgraphExecutor
          : transport.getSubgraphExecutor;
      if (!transportGetSubgraphExecutor) {
        throw new Error(
          `@graphql-mesh/transport-${kind} module does not export "getSubgraphExecutor"`,
        );
      }
      if (typeof transportGetSubgraphExecutor !== 'function') {
        throw new Error(
          `@graphql-mesh/transport-${kind} module's export "getSubgraphExecutor" is not a function`,
        );
      }
    }
    return function getSubgraphExecutor(opts) {
      return transportGetSubgraphExecutor({
        ...opts,
        transportEntry: {
          ...opts.transportEntry,
          options: {
            ...opts.transportEntry.options,
            ...transportOptions?.['*']?.[kind],
            ...transportOptions?.[opts.transportEntry.subgraph]?.[kind],
          },
        },
      });
    };
  };
}

function getTransportExecutor(
  transportExecutorFactoryGetter: TransportExecutorFactoryGetter,
  transportContext: TransportGetSubgraphExecutorOptions,
  transportExecutorStack: AsyncDisposableStack,
): MaybePromise<Executor> {
  const kind = transportContext.transportEntry?.kind || '';
  const subgraphName = transportContext.subgraphName || '';
  transportContext.logger?.info(`Loading transport ${kind} for subgraph ${subgraphName}`);
  return mapMaybePromise(transportExecutorFactoryGetter(kind), executor =>
    mapMaybePromise(executor(transportContext), executor => {
      if (isDisposable(executor)) {
        transportExecutorStack.use(executor);
      }
      return executor;
    }),
  );
}

/**
 * This function creates a executor factory that uses the transport packages,
 * and wraps them with the hooks
 */
export function getOnSubgraphExecute({
  onSubgraphExecuteHooks,
  transportContext,
  transportEntryMap,
  getSubgraphSchema,
  transportExecutorStack,
  transports,
  transportOptions,
}: {
  onSubgraphExecuteHooks: OnSubgraphExecuteHook[];
  transports?: Transports;
  transportOptions?: TransportOptions;
  transportContext?: TransportContext;
  transportEntryMap?: Record<string, TransportEntry>;
  getSubgraphSchema(subgraphName: string): GraphQLSchema;
  transportExecutorStack: AsyncDisposableStack;
}) {
  const subgraphExecutorMap = new Map<string, Executor>();
  const transportExecutorFactoryGetter = createTransportExecutorFactoryGetter(
    transports,
    transportOptions,
    transportContext.logger,
  );
  return function onSubgraphExecute(subgraphName: string, executionRequest: ExecutionRequest) {
    let executor: Executor = subgraphExecutorMap.get(subgraphName);
    // If the executor is not initialized yet, initialize it
    if (executor == null) {
      transportContext?.logger?.info(`Initializing executor for subgraph ${subgraphName}`);
      // Lazy executor that loads transport executor on demand
      executor = function lazyExecutor(subgraphExecReq: ExecutionRequest) {
        return mapMaybePromise(
          // Gets the transport executor for the given subgraph
          getTransportExecutor(
            transportExecutorFactoryGetter,
            transportContext
              ? {
                  ...transportContext,
                  subgraphName,
                  get subgraph() {
                    return getSubgraphSchema(subgraphName);
                  },
                  get transportEntry() {
                    return transportEntryMap?.[subgraphName];
                  },
                }
              : {
                  get subgraph() {
                    return getSubgraphSchema(subgraphName);
                  },
                  get transportEntry() {
                    return transportEntryMap?.[subgraphName];
                  },
                  subgraphName,
                },
            transportExecutorStack,
          ),
          executor_ => {
            // Wraps the transport executor with hooks
            executor = wrapExecutorWithHooks({
              executor: executor_,
              onSubgraphExecuteHooks,
              subgraphName,
              transportEntryMap,
              getSubgraphSchema,
            });
            // Caches the executor for future use
            subgraphExecutorMap.set(subgraphName, executor);
            return executor(subgraphExecReq);
          },
        );
      };
      // Caches the lazy executor to prevent race conditions
      subgraphExecutorMap.set(subgraphName, executor);
    }
    return executor(executionRequest);
  };
}

export interface WrapExecuteWithHooksOptions {
  executor: Executor;
  onSubgraphExecuteHooks: OnSubgraphExecuteHook[];
  subgraphName: string;
  transportEntryMap?: Record<string, TransportEntry>;
  getSubgraphSchema: (subgraphName: string) => GraphQLSchema;
}

/**
 * This function wraps the executor created by the transport package
 * with `onSubgraphExecuteHooks` to hook into the execution phase of subgraphs
 */
export function wrapExecutorWithHooks({
  executor,
  onSubgraphExecuteHooks,
  subgraphName,
  transportEntryMap,
  getSubgraphSchema,
}: WrapExecuteWithHooksOptions): Executor {
  if (onSubgraphExecuteHooks.length === 0) {
    return executor;
  }
  return function executorWithHooks(executionRequest: ExecutionRequest) {
    const onSubgraphExecuteDoneHooks: OnSubgraphExecuteDoneHook[] = [];
    return mapMaybePromise(
      iterateAsync(
        onSubgraphExecuteHooks,
        onSubgraphExecuteHook =>
          onSubgraphExecuteHook({
            get subgraph() {
              return getSubgraphSchema(subgraphName);
            },
            subgraphName,
            get transportEntry() {
              return transportEntryMap?.[subgraphName];
            },
            executionRequest,
            setExecutionRequest(newExecutionRequest) {
              executionRequest = newExecutionRequest;
            },
            executor,
            setExecutor(newExecutor) {
              executor = newExecutor;
            },
          }),
        onSubgraphExecuteDoneHooks,
      ),
      () => {
        if (onSubgraphExecuteDoneHooks.length === 0) {
          return executor(executionRequest);
        }
        return mapMaybePromise(executor(executionRequest), currentResult => {
          const executeDoneResults: OnSubgraphExecuteDoneResult[] = [];
          return mapMaybePromise(
            iterateAsync(
              onSubgraphExecuteDoneHooks,
              onSubgraphExecuteDoneHook =>
                onSubgraphExecuteDoneHook({
                  result: currentResult,
                  setResult(newResult: ExecutionResult) {
                    currentResult = newResult;
                  },
                }),
              executeDoneResults,
            ),
            () => {
              if (!isAsyncIterable(currentResult)) {
                return currentResult;
              }

              if (executeDoneResults.length === 0) {
                return currentResult;
              }

              const onNextHooks: OnSubgraphExecuteDoneResultOnNext[] = [];
              const onEndHooks: OnSubgraphExecuteDoneResultOnEnd[] = [];

              for (const executeDoneResult of executeDoneResults) {
                if (executeDoneResult.onNext) {
                  onNextHooks.push(executeDoneResult.onNext);
                }
                if (executeDoneResult.onEnd) {
                  onEndHooks.push(executeDoneResult.onEnd);
                }
              }

              if (onNextHooks.length === 0 && onEndHooks.length === 0) {
                return currentResult;
              }

              const asyncIterator = currentResult[Symbol.asyncIterator]();
              return mapAsyncIterator(
                asyncIterator,
                currentResult =>
                  mapMaybePromise(
                    iterateAsync(onNextHooks, onNext =>
                      onNext({
                        result: currentResult,
                        setResult: res => {
                          currentResult = res;
                        },
                      }),
                    ),
                    () => currentResult,
                  ),
                undefined,
                () =>
                  onEndHooks.length === 0 ? undefined : iterateAsync(onEndHooks, onEnd => onEnd()),
              );
            },
          );
        });
      },
    );
  };
}

export interface UnifiedGraphPlugin {
  onSubgraphExecute?: OnSubgraphExecuteHook;
}

export type OnSubgraphExecuteHook = (
  payload: OnSubgraphExecutePayload,
) => Promise<Maybe<OnSubgraphExecuteDoneHook | void>> | Maybe<OnSubgraphExecuteDoneHook | void>;

export interface OnSubgraphExecutePayload {
  subgraph: GraphQLSchema;
  subgraphName: string;
  transportEntry?: TransportEntry;
  executionRequest: ExecutionRequest;
  setExecutionRequest(executionRequest: ExecutionRequest): void;
  executor: Executor;
  setExecutor(executor: Executor): void;
}

export interface OnSubgraphExecuteDonePayload {
  result: AsyncIterable<ExecutionResult> | ExecutionResult;
  setResult(result: AsyncIterable<ExecutionResult> | ExecutionResult): void;
}

export type OnSubgraphExecuteDoneHook = (
  payload: OnSubgraphExecuteDonePayload,
) => MaybePromise<Maybe<OnSubgraphExecuteDoneResult | void>>;

export type OnSubgraphExecuteDoneResultOnNext = (
  payload: OnSubgraphExecuteDoneOnNextPayload,
) => MaybePromise<void>;

export interface OnSubgraphExecuteDoneOnNextPayload {
  result: ExecutionResult;
  setResult(result: ExecutionResult): void;
}

export type OnSubgraphExecuteDoneResultOnEnd = () => MaybePromise<void>;

export type OnSubgraphExecuteDoneResult = {
  onNext?: OnSubgraphExecuteDoneResultOnNext;
  onEnd?: OnSubgraphExecuteDoneResultOnEnd;
};

export function compareSchemas(
  a: DocumentNode | string | GraphQLSchema,
  b: DocumentNode | string | GraphQLSchema,
) {
  let aStr: string;
  if (typeof a === 'string') {
    aStr = a;
  } else if (isDocumentNode(a)) {
    aStr = print(a);
  } else {
    aStr = printSchemaWithDirectives(a);
  }
  let bStr: string;
  if (typeof b === 'string') {
    bStr = b;
  } else if (isDocumentNode(b)) {
    bStr = print(b);
  } else {
    bStr = printSchemaWithDirectives(b);
  }
  return aStr === bStr;
}

export function isDisposable(obj: any): obj is Disposable | AsyncDisposable {
  return obj?.[Symbol.dispose] != null || obj?.[Symbol.asyncDispose] != null;
}
