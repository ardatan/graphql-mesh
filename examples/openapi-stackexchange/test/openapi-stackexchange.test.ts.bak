import { readFile } from 'fs/promises';
import { join } from 'path';
import { findAndParseConfig } from '@graphql-mesh/cli';
import { getMesh, MeshInstance } from '@graphql-mesh/runtime';
import { printSchemaWithDirectives } from '@graphql-tools/utils';

describe.skip('Stack Exchange', () => {
  let mesh: MeshInstance;
  beforeAll(async () => {
    const config = await findAndParseConfig({
      dir: join(__dirname, '..'),
    });
    mesh = await getMesh(config);
  });
  afterAll(() => {
    mesh.destroy();
  });
  it('should generate correct schema', async () => {
    expect(printSchemaWithDirectives(mesh.schema)).toMatchSnapshot();
  });
  it('should return the correct data', async () => {
    const listQuestionsQuery = await readFile(
      join(__dirname, '..', 'list-questions.query.graphql'),
      'utf-8',
    );
    const result = await mesh.execute(listQuestionsQuery);
    expect(result).toMatchObject({
      data: {
        listUnansweredQuestions: {
          items: expect.arrayContaining([
            expect.objectContaining({
              title: expect.any(String),
              tags: expect.arrayContaining([expect.any(String)]),
              is_answered: expect.any(Boolean),
              answer_count: expect.any(Number),
              link: expect.any(String),
            }),
          ]),
        },
      },
    });
  });
});
