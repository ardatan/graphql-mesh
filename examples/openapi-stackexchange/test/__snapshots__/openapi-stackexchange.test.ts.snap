// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack Exchange should generate correct schema 1`] = `
"schema {
  query: Query
}

type Query {
  "Gets all the questions on the site.\\nThis method allows you make fairly flexible queries across the entire corpus of questions on a site."
  listQuestions(fromdate: Int, todate: Int, min: Int, max: Int, tagged: String, order: queryInput_listQuestions_order, sort: queryInput_listQuestions_sort, page: Int, pagesize: Int, site: String! = "stackoverflow"): QuestionsResponse
  "Gets all the questions on the site.\\nThis method allows you make fairly flexible queries across the entire corpus of questions on a site."
  listFeaturedQuestions(fromdate: Int, todate: Int, min: Int, max: Int, tagged: String, order: queryInput_listFeaturedQuestions_order, sort: queryInput_listFeaturedQuestions_sort, page: Int, pagesize: Int, site: String! = "stackoverflow"): QuestionsResponse
  "Returns questions which have received no answers.\\nCompare with /questions/unanswered which merely returns questions that the sites consider insufficiently well answered."
  listQuestionsWithoutAnswers(fromdate: Int, todate: Int, min: Int, max: Int, tagged: String, order: queryInput_listQuestionsWithoutAnswers_order, sort: queryInput_listQuestionsWithoutAnswers_sort, page: Int, pagesize: Int, site: String! = "stackoverflow"): QuestionsResponse
  "Gets all the questions on the site.\\nThis method allows you make fairly flexible queries across the entire corpus of questions on a site."
  listUnansweredQuestions(fromdate: Int, todate: Int, min: Int, max: Int, tagged: String, order: queryInput_listUnansweredQuestions_order, sort: queryInput_listUnansweredQuestions_sort, page: Int, pagesize: Int, site: String! = "stackoverflow"): QuestionsResponse
  "Returns all users on a site.\\n\\nThis method returns a list of users.\\n\\nThe sorts accepted by this method operate on the following fields of the user object:\\n\\nreputation – reputation\\ncreation – creation_date\\nname – display_name\\nmodified – last_modified_date\\n\\nreputation is the default sort.\\n\\nIt is possible to create moderately complex queries using sort, min, max, fromdate, and todate.\\nThe \`inname\` parameter lets consumers filter the results down to just those users with a certain substring in their display name. For example, \`inname=kevin\` will return all users with both users named simply \\"Kevin\\" or those with Kevin as one of (or part of) their names; such as \\"Kevin Montrose\\"."
  listUsers(inname: String, fromdate: Int, todate: Int, min: Int, max: Int, sort: queryInput_listUsers_sort, order: queryInput_listUsers_order, page: Int, pagesize: Int, site: String! = "stackoverflow"): UsersResponse
  "Returns the user associated with the passed access_token.\\n\\nThis method returns a [user](https://api.stackexchange.com/docs/types/user)."
  getMe(site: String! = "stackoverflow"): UsersResponse
  "Gets a subset of the reputation changes for users in {ids}.\\nReputation changes are intentionally scrubbed of some data to make it difficult to correlate votes on particular posts with user reputation changes. That being said, this method returns enough data for reasonable display of reputation trends.\\n{ids} can contain up to 100 semicolon delimited ids. To find ids programmatically look for user_id on user or shallow_user objects.\\nThis method returns a list of reputation objects."
  getUsersReputationChanges(userIds: String!, fromdate: Int, todate: Int, page: Int, pagesize: NonNegativeInt, site: String! = "stackoverflow"): ReputationResponse
  "Returns the reputation changed for the user associated with the given access_token.\\nThis method returns a list of [reputation changes](https://api.stackexchange.com/docs/types/reputation)."
  getMyReputation: ReputationResponse
  "Returns users' public reputation history.\\nThis method returns a list of reputation_history."
  getUsersReputationHistory(userIds: String!, page: Int, pagesize: Int, site: String! = "stackoverflow"): ReputationHistoryResponse
  "Returns user's public reputation history.\\n\\nThis method returns a list of [reputation_history](https://api.stackexchange.com/docs/types/reputation-history)."
  getMyReputationHistory(page: Int, pagesize: Int): ReputationHistoryResponse
  "Returns a list of [answers](https://api.stackexchange.com/docs/types/answer)."
  listAnswers(
    "This parameter can be the full domain name (ie. \\"stackoverflow.com\\"), or a short form identified by api_site_parameter on the site object."
    site: String
    "Unix epoch time"
    fromdate: Int
    "Unix epoch time"
    todate: Int
    "Unix epoch time"
    min: Int
    "Unix epoch time"
    max: Int
    sort: queryInput_listAnswers_sort
    order: queryInput_listAnswers_order
    page: Int
    pagesize: Int
  ): AnswersResponse
}

type QuestionsResponse {
  "A list of questions."
  items: [Question]
  has_more: Boolean
  quota_max: Int
  quota_remaining: Int
}

type Question {
  tags: [String]
  owner: QuestionOwner
  is_answered: Boolean
  view_count: Int
  answer_count: Int
  score: Int
  last_activity_date: Int
  creation_date: Int
  question_id: Int
  link: String
  title: String
}

type QuestionOwner {
  reputation: Int
  user_id: Int
  user_type: String
  profile_image: String
  display_name: String
  link: String
}

enum queryInput_listQuestions_order {
  desc
  asc
}

enum queryInput_listQuestions_sort {
  activity
  votes
  creation
  hot
  week
  month
}

enum queryInput_listFeaturedQuestions_order {
  desc
  asc
}

enum queryInput_listFeaturedQuestions_sort {
  activity
  votes
  creation
}

enum queryInput_listQuestionsWithoutAnswers_order {
  desc
  asc
}

enum queryInput_listQuestionsWithoutAnswers_sort {
  activity
  votes
  creation
}

enum queryInput_listUnansweredQuestions_order {
  desc
  asc
}

enum queryInput_listUnansweredQuestions_sort {
  activity
  votes
  creation
}

type UsersResponse {
  items: [User]
  has_more: Boolean
  quote_max: Int
  quota_remaining: Int
}

type User {
  badge_counts: UserBadgeCounts
  account_id: BigInt
  is_employee: Boolean
  last_modified_date: BigInt
  last_access_date: BigInt
  reputation_change_year: Int
  reputation_change_quarter: Int
  reputation_change_month: Int
  reputation_change_week: Int
  reputation_change_day: Int
  reputation: BigInt
  creation_date: BigInt
  user_type: String
  user_id: BigInt
  accept_rate: Int
  location: String
  website_url: URL
  link: URL
  profile_image: URL
  display_name: String
}

type UserBadgeCounts {
  bronze: Int
  silver: Int
  gold: Int
}

"The \`BigInt\` scalar type represents non-fractional signed whole numeric values."
scalar BigInt

"A field whose value conforms to the standard URL format as specified in RFC3986: https://www.ietf.org/rfc/rfc3986.txt."
scalar URL

enum queryInput_listUsers_sort {
  reputation
  creation
  name
  modified
}

enum queryInput_listUsers_order {
  desc
  asc
}

type ReputationResponse {
  "A list of reputation_history."
  items: [ReputationChange]
  has_more: Boolean
  quota_max: Int
  quota_remaining: Int
}

type ReputationChange {
  on_date: Int
  reputation_change: Int
  vote_type: query_getUsersReputationChanges_items_items_vote_type
  post_type: String
  post_id: Int
  user_id: Int
}

enum query_getUsersReputationChanges_items_items_vote_type {
  accepts
  bounties_won
  down_votes
  up_votes
}

"Integers that will have a value of 0 or more."
scalar NonNegativeInt

type ReputationHistoryResponse {
  "A list of reputation_history."
  items: [ReputationHistory]
  has_more: Boolean
  quota_max: Int
  quota_remaining: Int
}

type ReputationHistory {
  reputation_history_type: String
  reputation_change: Int
  post_id: Int
  creation_date: Int
  user_id: Int
}

type AnswersResponse {
  items: [Answer]
  has_more: Boolean
  backoff: Int
  quota_max: Int
  quota_remaining: Int
}

type Answer {
  accepted: Boolean
  answer_id: Int
  awarded_bounty_amount: Int
  awarded_bounty_users: [ShallowUser]
  body: String
  body_markdown: String
  can_flag: Boolean
  comment_count: Int
  comments: [Comment]
  community_owned_date: Int
  content_license: String
  creation_date: Int
  down_vote_count: Int
  is_accepted: Boolean
  last_activity_date: Int
  last_edit_date: Int
  last_editor: ShallowUser
  link: String
  locked_date: Int
  owner: ShallowUser
  question_id: Int
  score: Boolean
  share_link: Int
  tags: [String]
  title: String
  up_vote_count: Int
  upvoted: Boolean
}

type ShallowUser {
  id: String
}

type Comment {
  id: String
}

"\`activity\` corresponds to \`Answer. last_activity_date\` , \`creation\` to \`Answer.creation_date\` and \`votes\` to \`Answer.score\`"
enum queryInput_listAnswers_sort {
  activity
  creation
  votes
}

enum queryInput_listAnswers_order {
  desc
  asc
}"
`;
